### PR TITLE
Sync with Elasticsearch main branch

### PR DESCRIPTION
**Pull request by mashhurs**
### Description
This PR introduces changes aligned with Elasticsearch main branch which a big change, JDK 21 requirement.

### Author's checklist
- [x] Rebase after landing https://github.com/elastic/logstash-filter-elastic_integration/pull/170
- [x] Cut 8.x branch which will be compatible with 8.x stack versions
- [x] Run and check the build CI: https://buildkite.com/elastic/logstash-filter-elastic-integration-build/builds/392
  - validates the separated branches

### CODE REVIEW

File: .buildkite/build-pipeline.yml [hunk 1]
```diff
@@ -16,6 +16,7 @@ steps:
       SNAPSHOT: false
       INTEGRATION: true
       SECURE_INTEGRATION: true
+    branches: "8.x"
```
Comment (mashhurs) [inline .buildkite/build-pipeline.yml:19]: Test.
```suggestion
    branches: "mashhur-patch-1"
```

File: .buildkite/build-pipeline.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ steps:
       SNAPSHOT: false
       INTEGRATION: true
       SECURE_INTEGRATION: true
+    branches: "mashhur-patch-1"
```
Comment (mashhurs) [inline .buildkite/build-pipeline.yml:19]: After merging, rebasing #170, then cutting 8.x branch, change this.
```suggestion
    branches: "8x"
```

File: .buildkite/build-pipeline.yml [hunk 3]
```diff
@@ -8,18 +8,20 @@ agents:
...
+  - label: ":hammer: Build plugin with 8.x Elasticsearch version :elasticsearch:"
     command:
       - .buildkite/scripts/run_tests.sh
     env:
       ELASTIC_STACK_VERSION: "8.x"
+      ELASTICSEARCH_TREEISH: "8.x"
       SNAPSHOT: false
       INTEGRATION: true
       SECURE_INTEGRATION: true
+      TARGET_BRANCH: "mashhur-patch-1"
```
Comment (mashhurs) [inline .buildkite/build-pipeline.yml:20]:
```suggestion
      TARGET_BRANCH: "8.x"
```

File: build.gradle [hunk 4]
```diff
@@ -32,12 +32,10 @@ version                    "${file("VERSION").text.trim()}" // read from require
...
 // ===========================================================================
 
-java {
-    sourceCompatibility = 17
-    targetCompatibility = 17
-}
 tasks.withType(JavaCompile) {
     options.encoding = 'UTF-8'
+    sourceCompatibility = 21
+    targetCompatibility = 21
```
Comment (mashhurs) [inline build.gradle:38]: Self note: this will not work, needs to be changed with toolchain.

**Review by jsvd — APPROVED** [review]
LGTM

**Review by jsvd — COMMENTED** [review]
should we move gradle.properties ES TREEISH to main?

**Comment by mashhurs** [comment]
> should we move gradle.properties ES TREEISH to main?

Not necessarily from functional point of view as we are able to enforce the ES treeish with env var. If we need it for a visibility (make sure we don't make any mistake while developing), I am thinking to update the [GH PR template](https://github.com/elastic/logstash-filter-elastic_integration/blob/main/.github/PULL_REQUEST_TEMPLATE.md) which should guide 8.x and future release and probably I will also add a note to a plugin publisher.

**Comment by jsvd** [comment]
The issue I see is that, for a PR against the `main` branch of this plugin, for example, the main branch of ES is not being tested:

```
Resolved ELASTICSEARCH_TREEISH: 8.x
....
#31 34.81 > Task :downloadElasticsearchSourceZip
#31 34.81 Download https://github.com/elastic/elasticsearch/archive/refs/heads/8.x.zip
```

So we may need to force TREEISH to main in this branch, either through CI or the gradle.properties

**Comment by mashhurs** [comment]
> The issue I see is that, for a PR against the `main` branch of this plugin, for example, the main branch of ES is not being tested:
> 
> ```
> Resolved ELASTICSEARCH_TREEISH: 8.x
> ....
> #31 34.81 > Task :downloadElasticsearchSourceZip
> #31 34.81 Download https://github.com/elastic/elasticsearch/archive/refs/heads/8.x.zip
> ```
> 
> So we may need to force TREEISH to main in this branch, either through CI or the gradle.properties

I have [a PR](https://github.com/elastic/logstash-filter-elastic_integration/pull/177) to separate branches and give better names. How do you think if I add there? I have validated on my local to cover your concern if it is about failure.

- Unit test job
```
  - label: ":hammer: Unit tests with LS 8.x-SNAPSHOT & ES main :docker:"
    command:
      - .buildkite/scripts/run_tests.sh
    env:
      ELASTIC_STACK_VERSION: "8.x"
      INTEGRATION: false
      SNAPSHOT: true
      ELASTICSEARCH_TREEISH: main
```
Local result:
```
logstash-1  | Finished in 4.37 seconds (files took 3.5 seconds to load)
logstash-1  | 74 examples, 0 failures
logstash-1  | 
logstash-1  | Randomized with seed 56776
logstash-1  | 
logstash-1 exited with code 0
```

- Integration test job
```
  - label: ":hammer: Integration tests with LS 8.x-SNAPSHOT & ES main :docker:"
    command:
      - .buildkite/scripts/run_tests.sh
    env:
      ELASTIC_STACK_VERSION: "8.x"
      INTEGRATION: true
      LOG_LEVEL: "info"
      SECURE_INTEGRATION: true
      SNAPSHOT: true
      ELASTICSEARCH_TREEISH: main
```

Local result:
```log
logstash-1       |       raises an exception when condition is met
logstash-1       | 
logstash-1       | Finished in 14.59 seconds (files took 3.35 seconds to load)
logstash-1       | 49 examples, 0 failures
logstash-1       | 
logstash-1       | Randomized with seed 61530
logstash-1       | 
logstash-1 exited with code 0

```

**Comment by jsvd** [comment]
I'm fine with using #177 to improve the CI. LGTM

**Pull request merged**